src/smp/ipi.c

```diff
@@ -79,10 +78,45 @@ void ipiStallCoreCallback(bool_t irqPath)
     }
 }
 
+void ipi_wait(void)
+{
+    word_t cores = big_kernel_lock.totalCoreBarrier;
+    word_t *ipiSyncBarrier_count = &big_kernel_lock.ipiSyncBarrier_count;
+
+    if (__atomic_fetch_add(ipiSyncBarrier_count, 1, __ATOMIC_ACQ_REL) == cores) {
+        *ipiSyncBarrier_count = 0;
+        __atomic_thread_fence(__ATOMIC_RELEASE);
+    }
+
+    while (*ipiSyncBarrier_count != 0) {
```

Comment:
If the core1 resets *ipiSyncBarrier_count to 0, but before  core1 observes the change, another core3 starts another round of ipi_wait, the *ipiSyncBarrrier_count becomes greater than 0. Thus, core1 might stay in the while of the last round ipi_wait?

Reply:
> another core3 starts another round of ipi_wait

That's not possible I think. Only the core holding the kernel lock can initiate IPIs, and only cores receiving that IPI will call ipi_wait. IPIs are effectively serialised by the big kernel lock.

Anyway, if this would happen, it wouldn't work with the old code either, which did:
```
ipiSyncBarrier.globalsense = ~ipiSyncBarrier.globalsense;
```
and doing that twice also doesn't give a difference.

Reply:
the old code works since it does not wait on the counter.

Reply:
I'm not sure why the remote cores need to wait for all other cores to have handled the IPI instead of just returning to user space. Those `ip_wait` calls could probable be replaced with an `ipi_ack` instead. Perhaps it makes sense to split `ipi_wait` into `ipi_ack` and `ipi_wait`. To me it seems that the IPI sender only needs to wait for the other cores to ack the IPI. Now ipi_wait also works as a lock in case multiple CPUs want to do an IPI at the same time, but that seems unnecessary, as that's not possible anyway because of overwritten global function and args data.

Reply:
I don't see how it could have worked:
```
core0		core1		core2		core3
---------------------------------------------------------
sense=A
IPI for 1 & 2
count=1
		count=2
				count=3
				sense=~A
				count=0
						IPI for 1
						count=1
		count=2
		sense=A
		count=0
still sees A
```
The race window (if there is one) is smaller, but it's not gone.

Reply:
Sorry, I did not follow the example you gave. Maybe a SPIN model could confirm the new barrier code is correct.

Reply:
If sending blocking IPIs is not serialised by the big kernel lock, multiple IPI senders will re-use the same global data, and it will be broken, agreed? So in my example above core3 would not be able to send an IPI until core0 is done, so disregard that.

After my change there is a theoretical race window where:
```
core0		core1		core2
---------------------------------------------------------
take bkl
IPI for c1
		count=1
count=2
count=0
release bkl
				take bkl
				IPI for 1
				count=1
		still sees
		count != 0
		deadlock
```
But before my change that would have been:
```
sense=A

core0		core1		core2
---------------------------------------------------------
take bkl
IPI for c1
		count=1
count=2
sense=~A
count=0
release bkl
				take bkl
				IPI for 1
				count=1
		sees A != ~A
		no deadlock
```
However, if core 2 would send an IPI to core 3 and 3 would invert globalsense before core1 checked it, there is still a race, although no deadlock.

So I agree the original code was more robust against races. I'll have a look at how to fix this, maybe for now just add back the sense thing.

Reply:
> However, if core 2 would send an IPI to core 3 and 3 would invert globalsense before core1 checked it, there is still a race, although no deadlock.

Instead of a toggle, it could be an incrementing counter

Reply:
It'd probably be better to take out the ipi_wait in a separate PR so that the behavior change from this PR is smaller.

Reply:
> It'd probably be better to take out the ipi_wait in a separate PR so that the behavior change from this PR is smaller.

Agreed. Especially if that causes the IPC1001 failure (running the test in a loop here, but I can't reproduce it).